File: 3-LongestSubstringWithoutRepeatingCharacters/3-LongestSubstringWithoutRepeatingCharacters.cpp
// Last updated: 30/07/2026, 10:25:36
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        
        unordered_map<char, int> mp;
        int left = 0;
        int maxLength = 0;
        for (int right = 0; right < s.size(); right++) {
            mp[s[right]]++;
            while (mp[s[right]] > 1) {
                mp[s[left]]--;
                left++;
            }
            maxLength = max(maxLength, right - left + 1);
        }
        
        return maxLength;
    }
};

File: 3-LongestSubstringWithoutRepeatingCharacters/3-LongestSubstringWithoutRepeatingCharacters.js
// Last updated: 30/07/2026, 10:25:39
function lengthOfLongestSubstring(s) {
    const charIndex = new Map();
    let left = 0;
    let maxLen = 0;

    for (let right = 0; right < s.length; right++) {
        const char = s[right];

        // If char is in the window, shrink left past its last position
        if (charIndex.has(char) && charIndex.get(char) >= left) {
            left = charIndex.get(char) + 1;
        }

        charIndex.set(char, right);
        maxLen = Math.max(maxLen, right - left + 1);
    }

    return maxLen;
}